[dy] Add paystack source

dy46 opened the pull request:
# Summary

Add paystack source. Right now the only streams that are available are `customers` and `transactions`. 
<!-- Brief summary of what your code does -->

# Tests

tested locally with test customer and transaction data
<!-- How did you test your change? -->

cc: @wangxiaoyou1993 
<!-- Optionally mention someone to let them know about this pull request -->

wangxiaoyou1993 commented on mage_integrations/mage_integrations/sources/paystack/__init__.py:
```diff
@@ -0,0 +1,49 @@
+    def load_data(
+        self,
+        stream,
+        bookmarks: Dict = None,
```
Are you going to support bookmark?

wangxiaoyou1993 commented on mage_integrations/mage_integrations/sources/paystack/streams/base.py:
```diff
@@ -0,0 +1,172 @@
+
+        # Attempt to get the bookmark date from the state file (if one exists and is supplied).
+        LOGGER.info('Attempting to get the most recent bookmark_date for entity {}.'.format(self.ENTITY))
+        bookmark_date = get_last_record_value_for_table(self.state, table, 'bookmark_date')
```
I wonder why we don't pass the bookmark directly to this method. Why do we fetch bookmark again?

dy46 replied:
yeah sorry, i wanted to try to reuse the code in the chargebee sync since it's similar to paystack, so there might be some weird things like this. i'll try to polish it a bit more

wangxiaoyou1993 commented on mage_integrations/mage_integrations/sources/paystack/streams/base.py:
```diff
@@ -0,0 +1,172 @@
+            # this will make sure that bookmark does not go beyond (now - 2 minutes)
+            # so, no data will be missed due to API latency
+            max_date = min(max_date, to_date)
+            self.state = incorporate(
```
The bookmark property is "createdAt" in your TransactionsStream. But you use bookmark_date here, they have conflicts. 
The logic of updating state and bookmark values is here https://github.com/mage-ai/mage-ai/blob/master/mage_integrations/mage_integrations/sources/base.py#L432-L444

dy46 commented:
@wangxiaoyou1993 I updated the bookmark logic. I don't think I fully understood how the bookmark worked, but I think it makes more sense now. I also tested it with some test transactions that I added

wangxiaoyou1993 commented on mage_integrations/mage_integrations/sources/paystack/streams/base.py:
```diff
@@ -0,0 +1,166 @@
+from mage_integrations.sources.constants import REPLICATION_METHOD_INCREMENTAL
+from mage_integrations.sources.messages import write_schema
+from mage_integrations.sources.paystack.client import PaystackClient
+from mage_integrations.sources.state_utils import (
```
Remove unused imports. Also if `state_utils` is not shared, you don't need to move that file.

wangxiaoyou1993 commented on mage_integrations/mage_integrations/sources/paystack/streams/base.py:
```diff
@@ -0,0 +1,166 @@
+
+LOGGER = singer.get_logger()
+
+def is_selected(stream_catalog):
```
where do you use this method?

wangxiaoyou1993 commented on mage_integrations/mage_integrations/sources/paystack/streams/base.py:
```diff
@@ -0,0 +1,166 @@
+        self.client = client
+        self.substreams = []
+
+    def write_schema(self):
```
is it used?

wangxiaoyou1993 commented on mage_integrations/mage_integrations/sources/paystack/__init__.py:
```diff
@@ -0,0 +1,51 @@
+    def get_valid_replication_keys(self, stream_id):
+        return STREAMS[stream_id].VALID_REPLICATION_KEYS
+
+    def load_schemas_from_folder(self) -> Dict:
```
You might not need to override this method. I think it gets the same results with the base class's method: https://github.com/mage-ai/mage-ai/blob/master/mage_integrations/mage_integrations/sources/base.py#L659

wangxiaoyou1993 commented on mage_integrations/mage_integrations/sources/paystack/streams/base.py:
```diff
@@ -0,0 +1,166 @@
+    def get_class_path(self):
+        return os.path.dirname(inspect.getfile(self.__class__))
+
+    def load_schema_by_name(self, name):
```
If the `load_schemas_from_folder` in `paystack/__init__.py` is not needed, this method, `get_schema` and `get_class_path` are not needed as well.

wangxiaoyou1993 commented on mage_integrations/mage_integrations/sources/paystack/streams/base.py:
```diff
@@ -0,0 +1,166 @@
+import singer
+
+
+LOGGER = singer.get_logger()
```
Let's pass Paystack's logger to this class. `LOGGER` doesn't log the correct format that can be shown in Mage's logs page.

wangxiaoyou1993 commented on mage_integrations/mage_integrations/sources/paystack/streams/base.py:
```diff
@@ -0,0 +1,166 @@
+        # If there is no bookmark date, fall back to using the start date from the config file.
+        if bookmark_date is None:
+            LOGGER.info('Could not locate bookmark_date from STATE file. Falling back to start_date from config.json instead.')
+            bookmark_datetime = parse(self.config.get('start_date'))
```
will there always be a start_date?

wangxiaoyou1993 commented on mage_integrations/mage_integrations/sources/paystack/streams/base.py:
```diff
@@ -0,0 +1,166 @@
+        # If there is no bookmark date, fall back to using the start date from the config file.
+        if bookmark_date is None:
+            LOGGER.info('Could not locate bookmark_date from STATE file. Falling back to start_date from config.json instead.')
+            bookmark_datetime = parse(self.config.get('start_date'))
```
will there always be a start_date? can we have a default value?

dy46 replied:
i was planning to make it a required value in the config

dy46 replied:
ok i just added a default to be safe

dy46 commented:
@wangxiaoyou1993 addressed your comment

wangxiaoyou1993 commented on mage_integrations/mage_integrations/sources/chargebee/templates/config.json:
```diff
@@ -5,4 +5,3 @@
   "include_deleted": "True",
   "request_timeout": 300
 }
-  
```
@dy46 could you address this comment?

dy46 replied:
i added a new line, not sure why it's showing that it's not there

wangxiaoyou1993 commented on mage_integrations/mage_integrations/sources/paystack/streams/base.py:
```diff
@@ -0,0 +1,121 @@
+        bookmark_date = None
+
+        # Attempt to get the bookmark date from the state file (if one exists and is supplied).
+        self.logger.info('Attempting to get the most recent bookmark_date for entity {}.'.format(self.ENTITY))
```
nit: break the line if it's too long

wangxiaoyou1993 commented on mage_integrations/mage_integrations/sources/paystack/streams/base.py:
```diff
@@ -0,0 +1,121 @@
+
+        # If there is no bookmark date, fall back to using the start date from the config file.
+        if bookmark_date is None:
+            self.logger.info('Could not locate bookmark_date from STATE file. Falling back to start_date from config.json instead.')
```
same here. break the line.

wangxiaoyou1993 commented:
could you update the PR description `Tests` section?
